File: server/services/zenVectorService.py
#!/usr/bin/env python3
"""
ZenVector Agent - Advanced AI Agent with Vector Database Integration
Provides code similarity analysis, semantic search, and demographic data insights
"""

import json
import os
import uuid
import sys
from typing import List, Dict, Any, Optional, Tuple
from datetime import datetime
import logging

# Simple implementation without external dependencies for now
# TODO: Add chromadb, sentence-transformers when environment supports them

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class ZenVectorAgent:
    """
    ZenVector: Advanced AI Agent for Code Intelligence and Demographic Analysis
    
    Features:
    - Code similarity detection using vector embeddings
    - Semantic search across codebases
    - Demographic data pattern analysis
    - Multi-modal search capabilities
    """
    
    def __init__(self, db_path: str = "./vector_db"):
        """Initialize ZenVector Agent with Chroma vector database"""
        try:
            # Initialize Chroma client
            self.client = chromadb.PersistentClient(path=db_path)
            
            # Initialize sentence transformer for embeddings
            self.embedding_model = SentenceTransformer('all-MiniLM-L6-v2')
            
            # Create collections for different data types
            self.code_collection = self._get_or_create_collection("code_similarity")
            self.semantic_collection = self._get_or_create_collection("semantic_search")
            self.demographic_collection = self._get_or_create_collection("demographic_data")
            
            logger.info("ZenVector Agent initialized successfully")
            
        except Exception as e:
            logger.error(f"Failed to initialize ZenVector Agent: {e}")
            raise
    
    def _get_or_create_collection(self, name: str):
        """Get or create a Chroma collection"""
        try:
            return self.client.get_collection(name=name)
        except ValueError:
            return self.client.create_collection(name=name)
    
    def add_code_to_vector_db(self, project_id: str, code_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Add code snippets to vector database for similarity analysis
        
        Args:
            project_id: Unique project identifier
            code_data: Dictionary containing classes, methods, and code content
            
        Returns:
            Dictionary with processing results
        """
        try:
            processed_items = []
            
            for class_info in code_data.get('classes', []):
                # Process each class
                class_text = self._extract_class_features(class_info)
                class_id = f"{project_id}_{class_info['name']}"
                
                # Generate embedding
                embedding = self.embedding_model.encode(class_text).tolist()
                
                # Store in vector database
                self.code_collection.add(
                    embeddings=[embedding],
                    documents=[class_text],
                    metadatas=[{
                        'project_id': project_id,
                        'class_name': class_info['name'],
                        'type': class_info.get('type', 'unknown'),
                        'package': class_info.get('package', ''),
                        'methods_count': len(class_info.get('methods', [])),
                        'timestamp': datetime.now().isoformat()
                    }],
                    ids=[class_id]
                )
                
                processed_items.append({
                    'id': class_id,
                    'class_name': class_info['name'],
                    'embedding_size': len(embedding)
                })
            
            return {
                'status': 'success',
                'processed_items': len(processed_items),
                'items': processed_items
            }
            
        except Exception as e:
            logger.error(f"Error adding code to vector DB: {e}")
            return {'status': 'error', 'message': str(e)}
    
    def find_similar_code(self, query_code: str, project_id: Optional[str] = None, 
                         top_k: int = 5) -> List[Dict[str, Any]]:
        """
        Find similar code patterns using vector similarity search
        
        Args:
            query_code: Code snippet or description to search for
            project_id: Optional project filter
            top_k: Number of results to return
            
        Returns:
            List of similar code matches with similarity scores
        """
        try:
            # Generate embedding for query
            query_embedding = self.embedding_model.encode(query_code).tolist()
            
            # Build filter if project_id provided
            where_filter = {"project_id": {"$eq": project_id}} if project_id else None
            
            # Search for similar code
            results = self.code_collection.query(
                query_embeddings=[query_embedding],
                n_results=top_k,
                where=where_filter
            )
            
            # Format results
            similar_code = []
            for i, (doc, metadata, distance) in enumerate(zip(
                results['documents'][0],
                results['metadatas'][0], 
                results['distances'][0]
            )):
                similarity_score = 1 - distance  # Convert distance to similarity
                similar_code.append({
                    'rank': i + 1,
                    'class_name': metadata['class_name'],
                    'project_id': metadata['project_id'],
                    'type': metadata['type'],
                    'package': metadata['package'],
                    'similarity_score': round(similarity_score, 3),
                    'code_preview': doc[:200] + "..." if len(doc) > 200 else doc,
                    'methods_count': metadata.get('methods_count', 0)
                })
            
            return similar_code
            
        except Exception as e:
            logger.error(f"Error finding similar code: {e}")
            return []
    
    def semantic_search(self, query: str, search_type: str = "all", 
                       top_k: int = 10) -> Dict[str, Any]:
        """
        Perform semantic search across code and documentation
        
        Args:
            query: Natural language search query
            search_type: "code", "documentation", or "all"
            top_k: Number of results to return
            
        Returns:
            Dictionary with search results and metadata
        """
        try:
            query_embedding = self.embedding_model.encode(query).tolist()
            
            search_results = {
                'query': query,
                'search_type': search_type,
                'results': [],
                'total_found': 0
            }
            
            if search_type in ["code", "all"]:
                code_results = self.code_collection.query(
                    query_embeddings=[query_embedding],
                    n_results=top_k
                )
                
                for doc, metadata, distance in zip(
                    code_results['documents'][0],
                    code_results['metadatas'][0],
                    code_results['distances'][0]
                ):
                    search_results['results'].append({
                        'type': 'code',
                        'title': f"{metadata['class_name']} ({metadata['type']})",
                        'content': doc[:300] + "..." if len(doc) > 300 else doc,
                        'relevance_score': round(1 - distance, 3),
                        'metadata': metadata
                    })
            
            search_results['total_found'] = len(search_results['results'])
            
            # Sort by relevance score
            search_results['results'].sort(
                key=lambda x: x['relevance_score'], 
                reverse=True
            )
            
            return search_results
            
        except Exception as e:
            logger.error(f"Error in semantic search: {e}")
            return {'error': str(e), 'results': []}
    
    def analyze_demographic_patterns(self, demographic_data: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Analyze demographic data patterns using vector analysis
        
        Args:
            demographic_data: List of demographic records
            
        Returns:
            Analysis results with patterns and insights
        """
        try:
            if not demographic_data:
                return {'error': 'No demographic data provided'}
            
            # Convert to DataFrame for analysis
            df = pd.DataFrame(demographic_data)
            
            # Generate text representations for each record
            demographic_texts = []
            for record in demographic_data:
                text = self._create_demographic_text(record)
                demographic_texts.append(text)
            
            # Generate embeddings
            embeddings = self.embedding_model.encode(demographic_texts)
            
            # Store in demographic collection
            record_ids = [str(uuid.uuid4()) for _ in demographic_data]
            
            self.demographic_collection.add(
                embeddings=embeddings.tolist(),
                documents=demographic_texts,
                metadatas=[{
                    'record_type': 'demographic',
                    'timestamp': datetime.now().isoformat(),
                    **{k: str(v) for k, v in record.items()}
                } for record in demographic_data],
                ids=record_ids
            )
            
            # Perform clustering analysis
            analysis_results = self._analyze_demographic_clusters(embeddings, df)
            
            return {
                'status': 'success',
                'records_processed': len(demographic_data),
                'analysis': analysis_results,
                'patterns_found': len(analysis_results.get('clusters', []))
            }
            
        except Exception as e:
            logger.error(f"Error analyzing demographic patterns: {e}")
            return {'error': str(e)}
    
    def search_demographic_data(self, query: str, top_k: int = 10) -> List[Dict[str, Any]]:
        """
        Search demographic data using natural language queries
        
        Args:
            query: Natural language search query
            top_k: Number of results to return
            
        Returns:
            List of matching demographic records
        """
        try:
            query_embedding = self.embedding_model.encode(query).tolist()
            
            results = self.demographic_collection.query(
                query_embeddings=[query_embedding],
                n_results=top_k
            )
            
            demographic_matches = []
            for doc, metadata, distance in zip(
                results['documents'][0],
                results['metadatas'][0],
                results['distances'][0]
            ):
                demographic_matches.append({
                    'content': doc,
                    'relevance_score': round(1 - distance, 3),
                    'metadata': {k: v for k, v in metadata.items() 
                              if not k.startswith('_')},
                    'match_type': 'demographic_data'
                })
            
            return demographic_matches
            
        except Exception as e:
            logger.error(f"Error searching demographic data: {e}")
            return []
    
    def get_agent_statistics(self) -> Dict[str, Any]:
        """Get ZenVector agent usage statistics"""
        try:
            stats = {
                'agent_name': 'ZenVector',
                'collections': {
                    'code_similarity': self.code_collection.count(),
                    'semantic_search': self.semantic_collection.count(),
                    'demographic_data': self.demographic_collection.count()
                },
                'total_vectors': 0,
                'capabilities': [
                    'Code Similarity Detection',
                    'Semantic Code Search',
                    'Demographic Data Analysis',
                    'Pattern Recognition',
                    'Multi-modal Search'
                ],
                'embedding_model': 'all-MiniLM-L6-v2',
                'vector_database': 'ChromaDB'
            }
            
            stats['total_vectors'] = sum(stats['collections'].values())
            
            return stats
            
        except Exception as e:
            logger.error(f"Error getting agent statistics: {e}")
            return {'error': str(e)}
    
    def _extract_class_features(self, class_info: Dict[str, Any]) -> str:
        """Extract textual features from class information"""
        features = []
        
        # Class name and type
        features.append(f"Class: {class_info['name']}")
        features.append(f"Type: {class_info.get('type', 'unknown')}")
        
        # Package information
        if class_info.get('package'):
            features.append(f"Package: {class_info['package']}")
        
        # Annotations
        if class_info.get('annotations'):
            features.append(f"Annotations: {', '.join(class_info['annotations'])}")
        
        # Methods
        methods = class_info.get('methods', [])
        if methods:
            method_names = [method.get('name', '') for method in methods[:5]]  # First 5 methods
            features.append(f"Methods: {', '.join(method_names)}")
        
        # Fields
        fields = class_info.get('fields', [])
        if fields:
            field_names = [field.get('name', '') for field in fields[:5]]  # First 5 fields
            features.append(f"Fields: {', '.join(field_names)}")
        
        return ' | '.join(features)
    
    def _create_demographic_text(self, record: Dict[str, Any]) -> str:
        """Create text representation of demographic record"""
        text_parts = []
        
        for key, value in record.items():
            if value is not None and str(value).strip():
                text_parts.append(f"{key}: {value}")
        
        return ' | '.join(text_parts)
    
    def _analyze_demographic_clusters(self, embeddings: np.ndarray, 
                                    df: pd.DataFrame) -> Dict[str, Any]:
        """Analyze demographic data clusters"""
        try:
            from sklearn.cluster import KMeans
            from sklearn.metrics import silhouette_score
            
            # Determine optimal number of clusters
            n_samples = len(embeddings)
            max_clusters = min(8, n_samples - 1) if n_samples > 1 else 1
            
            if max_clusters < 2:
                return {'clusters': [], 'analysis': 'Insufficient data for clustering'}
            
            # Perform K-means clustering
            kmeans = KMeans(n_clusters=max_clusters, random_state=42)
            cluster_labels = kmeans.fit_predict(embeddings)
            
            # Calculate silhouette score
            silhouette_avg = silhouette_score(embeddings, cluster_labels)
            
            # Analyze clusters
            clusters = []
            for i in range(max_clusters):
                cluster_indices = np.where(cluster_labels == i)[0]
                cluster_data = df.iloc[cluster_indices]
                
                clusters.append({
                    'cluster_id': i,
                    'size': len(cluster_indices),
                    'percentage': round(len(cluster_indices) / n_samples * 100, 1),
                    'characteristics': self._describe_cluster(cluster_data)
                })
            
            return {
                'clusters': clusters,
                'silhouette_score': round(silhouette_avg, 3),
                'total_clusters': max_clusters,
                'analysis': 'Demographic clustering completed successfully'
            }
            
        except Exception as e:
            logger.error(f"Error in demographic clustering: {e}")
            return {'clusters': [], 'analysis': f'Clustering failed: {str(e)}'}
    
    def _describe_cluster(self, cluster_data: pd.DataFrame) -> Dict[str, Any]:
        """Describe characteristics of a demographic cluster"""
        description = {}
        
        for column in cluster_data.columns:
            if cluster_data[column].dtype == 'object':
                # Categorical data
                mode_value = cluster_data[column].mode()
                if not mode_value.empty:
                    description[f"most_common_{column}"] = mode_value.iloc[0]
            else:
                # Numerical data
                description[f"avg_{column}"] = round(cluster_data[column].mean(), 2)
        
        return description

# Initialize global ZenVector agent instance
zen_vector_agent = None

def get_zen_vector_agent():
    """Get global ZenVector agent instance"""
    global zen_vector_agent
    if zen_vector_agent is None:
        zen_vector_agent = ZenVectorAgent()
    return zen_vector_agent

if __name__ == "__main__":
    # Test the ZenVector agent
    agent = ZenVectorAgent()
    stats = agent.get_agent_statistics()
    print("ZenVector Agent initialized successfully!")
    print(f"Agent Stats: {json.dumps(stats, indent=2)}")

File: server/services/zenVectorService.ts
/**
 * ZenVector Agent - Advanced AI Agent with Vector Database Integration
 * Provides code similarity analysis, semantic search, and demographic data insights
 * TypeScript implementation for immediate integration
 */

import OpenAI from 'openai';
import fs from 'fs/promises';
import path from 'path';

interface CodeSimilarityMatch {
  rank: number;
  className: string;
  projectId: string;
  type: string;
  package: string;
  similarityScore: number;
  codePreview: string;
  methodsCount: number;
}

interface SemanticSearchResult {
  query: string;
  searchType: string;
  results: Array<{
    type: string;
    title: string;
    content: string;
    relevanceScore: number;
    metadata: any;
  }>;
  totalFound: number;
}

interface DemographicPattern {
  clusterId: number;
  size: number;
  percentage: number;
  characteristics: Record<string, any>;
}

interface ZenVectorStats {
  agentName: string;
  collections: {
    codeSimilarity: number;
    semanticSearch: number;
    demographicData: number;
  };
  totalVectors: number;
  capabilities: string[];
  embeddingModel: string;
  vectorDatabase: string;
}

class ZenVectorAgent {
  private openai: OpenAI | null = null;
  private codeDatabase: Map<string, any> = new Map();
  private semanticDatabase: Map<string, any> = new Map();
  private demographicDatabase: Map<string, any> = new Map();
  private dbPath: string;

  constructor(dbPath: string = './vector_db') {
    this.dbPath = dbPath;
    
    // Initialize OpenAI if API key is available
    if (process.env.OPENAI_API_KEY) {
      this.openai = new OpenAI({
        apiKey: process.env.OPENAI_API_KEY,
      });
    }

    console.log('ZenVector Agent initialized successfully');
  }

  /**
   * Add code to vector database for similarity analysis
   */
  async addCodeToVectorDb(projectId: string, codeData: any): Promise<any> {
    try {
      const processedItems: any[] = [];

      for (const classInfo of codeData.classes || []) {
        const classText = this.extractClassFeatures(classInfo);
        const classId = `${projectId}_${classInfo.name}`;
        
        // Generate simple text embedding representation
        const embedding = await this.generateTextEmbedding(classText);
        
        // Store in in-memory database
        this.codeDatabase.set(classId, {
          id: classId,
          projectId,
          className: classInfo.name,
          type: classInfo.type || 'unknown',
          package: classInfo.package || '',
          methodsCount: classInfo.methods?.length || 0,
          content: classText,
          embedding,
          timestamp: new Date().toISOString()
        });

        processedItems.push({
          id: classId,
          className: classInfo.name,
          embeddingSize: embedding.length
        });
      }

      return {
        status: 'success',
        processedItems: processedItems.length,
        items: processedItems
      };

    } catch (error) {
      console.error('Error adding code to vector DB:', error);
      return { status: 'error', message: error.message };
    }
  }

  /**
   * Find similar code patterns using text similarity
   */
  async findSimilarCode(queryCode: string, projectId?: string, topK: number = 5): Promise<CodeSimilarityMatch[]> {
    try {
      const queryEmbedding = await this.generateTextEmbedding(queryCode);
      const similarities: Array<{ item: any; score: number }> = [];

      // Calculate similarities with stored code
      for (const [id, item] of this.codeDatabase.entries()) {
        if (projectId && item.projectId !== projectId) continue;

        const similarity = this.calculateSimilarity(queryEmbedding, item.embedding);
        similarities.push({ item, score: similarity });
      }

      // Sort by similarity score and take top K
      similarities.sort((a, b) => b.score - a.score);
      const topResults = similarities.slice(0, topK);

      return topResults.map((result, index) => ({
        rank: index + 1,
        className: result.item.className,
        projectId: result.item.projectId,
        type: result.item.type,
        package: result.item.package,
        similarityScore: Math.round(result.score * 1000) / 1000,
        codePreview: result.item.content.substring(0, 200) + "...",
        methodsCount: result.item.methodsCount
      }));

    } catch (error) {
      console.error('Error finding similar code:', error);
      return [];
    }
  }

  /**
   * Perform semantic search across code and documentation
   */
  async semanticSearch(query: string, searchType: string = "all", topK: number = 10): Promise<SemanticSearchResult> {
    try {
      const queryEmbedding = await this.generateTextEmbedding(query);
      const searchResults: SemanticSearchResult = {
        query,
        searchType,
        results: [],
        totalFound: 0
      };

      if (searchType === "code" || searchType === "all") {
        const similarities: Array<{ item: any; score: number }> = [];

        for (const [id, item] of this.codeDatabase.entries()) {
          const similarity = this.calculateSimilarity(queryEmbedding, item.embedding);
          similarities.push({ item, score: similarity });
        }

        similarities.sort((a, b) => b.score - a.score);
        const topResults = similarities.slice(0, topK);

        for (const result of topResults) {
          searchResults.results.push({
            type: 'code',
            title: `${result.item.className} (${result.item.type})`,
            content: result.item.content.substring(0, 300) + "...",
            relevanceScore: Math.round(result.score * 1000) / 1000,
            metadata: {
              projectId: result.item.projectId,
              className: result.item.className,
              type: result.item.type,
              package: result.item.package
            }
          });
        }
      }

      searchResults.totalFound = searchResults.results.length;
      searchResults.results.sort((a, b) => b.relevanceScore - a.relevanceScore);

      return searchResults;

    } catch (error) {
      console.error('Error in semantic search:', error);
      return { query, searchType, results: [], totalFound: 0 };
    }
  }

  /**
   * Analyze demographic data patterns
   */
  async analyzeDemographicPatterns(demographicData: any[]): Promise<any> {
    try {
      if (!demographicData || demographicData.length === 0) {
        return { error: 'No demographic data provided' };
      }

      // Store demographic data
      const recordIds: string[] = [];
      for (const record of demographicData) {
        const recordId = `demo_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
        const text = this.createDemographicText(record);
        const embedding = await this.generateTextEmbedding(text);

        this.demographicDatabase.set(recordId, {
          id: recordId,
          content: text,
          embedding,
          metadata: {
            recordType: 'demographic',
            timestamp: new Date().toISOString(),
            ...record
          }
        });

        recordIds.push(recordId);
      }

      // Perform simple clustering analysis
      const patterns = this.analyzeDemographicClusters(demographicData);

      return {
        status: 'success',
        recordsProcessed: demographicData.length,
        analysis: patterns,
        patternsFound: patterns.clusters?.length || 0
      };

    } catch (error) {
      console.error('Error analyzing demographic patterns:', error);
      return { error: error.message };
    }
  }

  /**
   * Search demographic data using natural language queries
   */
  async searchDemographicData(query: string, topK: number = 10): Promise<any[]> {
    try {
      const queryEmbedding = await this.generateTextEmbedding(query);
      const similarities: Array<{ item: any; score: number }> = [];

      for (const [id, item] of this.demographicDatabase.entries()) {
        const similarity = this.calculateSimilarity(queryEmbedding, item.embedding);
        similarities.push({ item, score: similarity });
      }

      similarities.sort((a, b) => b.score - a.score);
      const topResults = similarities.slice(0, topK);

      return topResults.map(result => ({
        content: result.item.content,
        relevanceScore: Math.round(result.score * 1000) / 1000,
        metadata: result.item.metadata,
        matchType: 'demographic_data'
      }));

    } catch (error) {
      console.error('Error searching demographic data:', error);
      return [];
    }
  }

  /**
   * Get ZenVector agent usage statistics
   */
  getAgentStatistics(): ZenVectorStats {
    return {
      agentName: 'ZenVector',
      collections: {
        codeSimilarity: this.codeDatabase.size,
        semanticSearch: this.semanticDatabase.size,
        demographicData: this.demographicDatabase.size
      },
      totalVectors: this.codeDatabase.size + this.semanticDatabase.size + this.demographicDatabase.size,
      capabilities: [
        'Code Similarity Detection',
        'Semantic Code Search',
        'Demographic Data Analysis',
        'Pattern Recognition',
        'Multi-modal Search'
      ],
      embeddingModel: 'OpenAI text-embedding-ada-002',
      vectorDatabase: 'In-Memory Vector Store'
    };
  }

  private extractClassFeatures(classInfo: any): string {
    const features: string[] = [];

    // Class name and type
    features.push(`Class: ${classInfo.name}`);
    features.push(`Type: ${classInfo.type || 'unknown'}`);

    // Package information
    if (classInfo.package) {
      features.push(`Package: ${classInfo.package}`);
    }

    // Annotations
    if (classInfo.annotations?.length > 0) {
      features.push(`Annotations: ${classInfo.annotations.join(', ')}`);
    }

    // Methods
    const methods = classInfo.methods || [];
    if (methods.length > 0) {
      const methodNames = methods.slice(0, 5).map((method: any) => method.name || '');
      features.push(`Methods: ${methodNames.join(', ')}`);
    }

    // Fields
    const fields = classInfo.fields || [];
    if (fields.length > 0) {
      const fieldNames = fields.slice(0, 5).map((field: any) => field.name || '');
      features.push(`Fields: ${fieldNames.join(', ')}`);
    }

    return features.join(' | ');
  }

  private createDemographicText(record: any): string {
    const textParts: string[] = [];

    for (const [key, value] of Object.entries(record)) {
      if (value != null && String(value).trim()) {
        textParts.push(`${key}: ${value}`);
      }
    }

    return textParts.join(' | ');
  }

  private async generateTextEmbedding(text: string): Promise<number[]> {
    try {
      if (this.openai) {
        // Use OpenAI embeddings for better accuracy
        const response = await this.openai.embeddings.create({
          model: "text-embedding-ada-002",
          input: text,
        });
        return response.data[0].embedding;
      } else {
        // Simple fallback: character frequency vector
        return this.generateSimpleEmbedding(text);
      }
    } catch (error) {
      console.warn('OpenAI embedding failed, using simple embedding:', error.message);
      return this.generateSimpleEmbedding(text);
    }
  }

  private generateSimpleEmbedding(text: string): number[] {
    // Simple character frequency-based embedding
    const charCounts = new Array(256).fill(0);
    for (let i = 0; i < text.length; i++) {
      const charCode = text.charCodeAt(i);
      if (charCode < 256) {
        charCounts[charCode]++;
      }
    }

    // Normalize
    const total = charCounts.reduce((sum, count) => sum + count, 0);
    return charCounts.map(count => total > 0 ? count / total : 0);
  }

  private calculateSimilarity(embedding1: number[], embedding2: number[]): number {
    // Cosine similarity
    let dotProduct = 0;
    let norm1 = 0;
    let norm2 = 0;

    const minLength = Math.min(embedding1.length, embedding2.length);

    for (let i = 0; i < minLength; i++) {
      dotProduct += embedding1[i] * embedding2[i];
      norm1 += embedding1[i] * embedding1[i];
      norm2 += embedding2[i] * embedding2[i];
    }

    const magnitude = Math.sqrt(norm1) * Math.sqrt(norm2);
    return magnitude > 0 ? dotProduct / magnitude : 0;
  }

  private analyzeDemographicClusters(data: any[]): any {
    // Simple clustering analysis
    const clusters: DemographicPattern[] = [];
    
    // Group by common attributes (simplified)
    const groupedData = new Map<string, any[]>();
    
    for (const record of data) {
      // Create a simple grouping key based on some attributes
      const keys = Object.keys(record).slice(0, 3); // First 3 attributes
      const groupKey = keys.map(key => `${key}:${record[key]}`).join('|');
      
      if (!groupedData.has(groupKey)) {
        groupedData.set(groupKey, []);
      }
      groupedData.get(groupKey)!.push(record);
    }

    let clusterId = 0;
    for (const [key, group] of groupedData.entries()) {
      clusters.push({
        clusterId: clusterId++,
        size: group.length,
        percentage: Math.round((group.length / data.length) * 100 * 10) / 10,
        characteristics: this.describeCluster(group)
      });
    }

    return {
      clusters,
      totalClusters: clusters.length,
      analysis: 'Demographic clustering completed successfully'
    };
  }

  private describeCluster(clusterData: any[]): Record<string, any> {
    const description: Record<string, any> = {};
    
    if (clusterData.length === 0) return description;

    // Get all unique keys
    const allKeys = new Set<string>();
    clusterData.forEach(record => {
      Object.keys(record).forEach(key => allKeys.add(key));
    });

    // Analyze each attribute
    for (const key of allKeys) {
      const values = clusterData.map(record => record[key]).filter(val => val != null);
      
      if (values.length > 0) {
        if (typeof values[0] === 'string') {
          // Most common string value
          const valueCounts = new Map<string, number>();
          values.forEach(val => {
            valueCounts.set(val, (valueCounts.get(val) || 0) + 1);
          });
          
          let mostCommon = '';
          let maxCount = 0;
          for (const [val, count] of valueCounts.entries()) {
            if (count > maxCount) {
              maxCount = count;
              mostCommon = val;
            }
          }
          description[`most_common_${key}`] = mostCommon;
        } else if (typeof values[0] === 'number') {
          // Average for numeric values
          const sum = values.reduce((acc, val) => acc + Number(val), 0);
          description[`avg_${key}`] = Math.round(sum / values.length * 100) / 100;
        }
      }
    }

    return description;
  }
}

// Global instance
let zenVectorAgent: ZenVectorAgent | null = null;

export function getZenVectorAgent(): ZenVectorAgent {
  if (!zenVectorAgent) {
    zenVectorAgent = new ZenVectorAgent();
  }
  return zenVectorAgent;
}

export default ZenVectorAgent;
export type { CodeSimilarityMatch, SemanticSearchResult, DemographicPattern, ZenVectorStats };